README.md
# Bitcoin Price Notification

# Description

This is a python package which is used to give regular notifications on different platforms like  email,telegram and IFTTT app notifications. 

Main aim is to give notification on selected platform on regular intervals which can be set according to the user, by default which is 1 minute and also user can set threshold price which is set to ₹4,00,000 by default.


![Image of Bitcoin](https://cdn.pixabay.com/photo/2016/11/10/05/09/bitcoin-1813503_960_720.jpg)

## Installation

Use the package manager [pip](https://pip.pypa.io/en/stable/) to install bitcoin notification package.

```bash
pip install bitcoin-notifier-rajatbhatt500
```

## Prerequisite
To use this package one should have below things:

* install the IFTTT App to receive the emergency notifications
* An Email account to get notification.
* Telegram App and must join this channel.click the [Telegram channel](https://t.me/bitcoin_notificationrajat) to join the channel.

## Usage
For Help Menu

```bash
bitcoin-notification --help
```

You'll see below message in command line
```bash
usage:             Usage: This app gives the price of 1 Bitcoin in INR.
            Destination(-d) must be provided else destination will be
            telegram by default.
            To recive notification from IFTTT install IFTTT mobile app.
            To recive notifications on Telegram install Telegram app
            and join this channel https://t.me/bitcoin_notificationrajat .

            Prerequisite : INSTALL IFTTT APP AND TELEGRAM APP
            and TO RECIVE NOTIFICATON.
            PRESS Ctrl+C to terminate the app


BITCOIN PRICE NOTIFIER

optional arguments:
  -h, --help            show this help message and exit
  -i interval, --interval interval
                        Time interval in minutes
  -t threshold, --threshold threshold
                        Threshold in INR(₹)
  -d destnation, --destination destnation
                        There are various options to recive notificationsfrom us (1)IFTTT app (2) Telegram app (3)
                        Email

Copyright @Rajat Bhatt
```
To here the example for notification in telegram:
```bash
bitcoin-notification -t 600000 -i 2 -d telegram
```

## License
[MIT](https://choosealicense.com/licenses/mit/)

bitcoin-notification/CHANGELOG.txt
Change Log
------------

0.0.4 (16/09/2020)
------------------------
Fourth Release

bitcoin-price-notifier/main.py

import requests
import time
import argparse
import sys
from requests.exceptions import RequestException, Timeout, TooManyRedirects
import re


class BitcoinNotification:
    def __init__(self, Bitcoin_Api_URL, Webhooks_URL):
        self.Bitcoin_Api_URL = Bitcoin_Api_URL
        self.Webhooks_URL = Webhooks_URL

    def getBitcoinPrice(self):

        try:
            api_url = self.Bitcoin_Api_URL
            # Api calling for bitcoin prices in INR
            response = requests.get(api_url)
            json_response = response.json()
            currentBitcoinPrice = json_response['bpi']['INR']['rate_float']
        # print cutomized msg if there is no internet access
        except RequestException:
            print(' NO Internet!')
            print('Please check your Internet'
                  ' connection and run program again.')
            sys.exit()
        except (ConnectionError, Timeout, TooManyRedirects) as e:
            print(e)
        # returning bitcoin price
        return round(currentBitcoinPrice, 2)

    def postWebhooks(self, event, value):

        ifttt_event_url = self.Webhooks_URL.format(event)
        requests.post(ifttt_event_url, json=value)

    def sendIFTTTEmergencyNotificaton(self, time_interval):

        try:
            price = self.getBitcoinPrice()
            value = {'value1': price}
            self.postWebhooks('bitcoin_price_emergency', value)
            print("Emergency notification sent")
            time.sleep(float(time_interval[0]) * 60)

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')

    def sendTelegramNotificaton(self, time_interval):

        try:
            price = self.getBitcoinPrice()
            value = {'value1': price}
            self.postWebhooks('bitcoin_price_update', value)
            print("Telegram Notification sent")
            time.sleep(float(time_interval[0]) * 60)

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')

    def sendGmailNotification(self, time_interval):

        try:
            regex = '^[a-z0-9]+[\._]?[a-z0-9]+[@]\w+[.]\w{2,3}$'
            name = input('Please enter your name: ')
            email = input('Please enter your Email: ')
            # chceking email is valid
            if(re.search(regex, email)):
                price = self.getBitcoinPrice()
                value = {'value1': email, 'value2': name, 'value3': price}
                self.postWebhooks('email_bitcoin_notification', value)
                print('Email notification sent')
                time.sleep(float(time_interval[0]) * 60)
            else:
                print('Invalid Email ,'+name)
                print('Please re-run the program and provide valid email')
                sys.exit()

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')


# driver code
if __name__ == "__main__":

    Bitcoin_Api_URL = 'https://api.coindesk.com/v1/bpi/currentprice/INR.json'
    # for bitcoin.notification123@gmail.com
    Webhooks_URL = 'https://maker.ifttt.com/trigger/{}/with/key/jrvNIRNL-vboEF2g3mcaDoUitHgy-Z180veQ4JkbYNf'

    b1 = BitcoinNotification(Bitcoin_Api_URL, Webhooks_URL)
    current_bitcoinprice = b1.getBitcoinPrice()
    parser = argparse.ArgumentParser(
        usage='''\
            Usage: This app gives the price of 1 Bitcoin in INR.
            Destination(-d) must be provided else destination will be
            telegram by default.
            To recive notification from IFTTT install IFTTT mobile app.
            To recive notifications on Telegram install Telegram app
            and join this channel @IFTTT.

            Prerequisite : INSTALL IFTTT APP AND TELEGRAM APP
            and TO RECIVE NOTIFICATON.
            PRESS Ctrl+C to terminate the app
        ''',
        description="BITCOIN PRICE NOTIFIER ",
        epilog="Copyright @Rajat Bhatt")
    # command line variable for time gap
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        nargs=1,
        metavar="interval",
        default=[1],
        help="Time interval in minutes")
    # command line variable for threshold
    parser.add_argument(
        "-t",
        "--threshold",
        type=int,
        nargs=1,
        metavar="threshold",
        default=[400000],
        help="Threshold in INR(₹)")

    # command line variable for destination
    parser.add_argument(
        "-d",
        "--destination",
        metavar='destnation',
        default='telegram',
        help='There are various options to recive notifications'
        'from us (1)IFTTT app (2) Telegram app (3) Email')

    new_args = parser.parse_args()
    print('Running Application with time interval of ', new_args.interval[0],
          'mins and threshold = ₹', new_args.threshold[0], 'and Destination= ',
          new_args.destination)

    # calls the function to send notifications
    if (new_args.destination == 'telegram' and
            new_args.threshold[0] < current_bitcoinprice):
        print('''\
            To recive the notification
        from Telegram, install the telegram app and join the
        channel using this link - https://t.me/bitcoin_notificationrajat or
        follow @Bitcoinprice_notificationrajat
        ''')
        b1.sendTelegramNotificaton(new_args.interval)

    if (new_args.destination == 'ifttt' or
            new_args.threshold[0] > current_bitcoinprice):

        b1.sendIFTTTEmergencyNotificaton(new_args.interval)

    if (new_args.destination == 'email'):
        b1.sendGmailNotification(new_args.interval)


bitcoin-notification/bitcoin_notifier_rajatbhatt500/bitcoin_price.py

import requests
import time
import argparse
import sys
from requests.exceptions import RequestException, Timeout, TooManyRedirects
import re


class BitcoinNotification:
    def __init__(self, Bitcoin_Api_URL, Webhooks_URL):
        self.Bitcoin_Api_URL = Bitcoin_Api_URL
        self.Webhooks_URL = Webhooks_URL

    def getBitcoinPrice(self):

        try:
            api_url = self.Bitcoin_Api_URL
            # Api calling for bitcoin prices in INR
            response = requests.get(api_url)
            json_response = response.json()
            currentBitcoinPrice = json_response['bpi']['INR']['rate_float']
        # print cutomized msg if there is no internet access
        except RequestException:
            print(' NO Internet!')
            print('Please check your Internet'
                  ' connection and run program again.')
            sys.exit()
        except (ConnectionError, Timeout, TooManyRedirects) as e:
            print(e)
        # returning bitcoin price
        return round(currentBitcoinPrice, 2)

    def postWebhooks(self, event, value):

        ifttt_event_url = self.Webhooks_URL.format(event)
        requests.post(ifttt_event_url, json=value)

    def sendIFTTTEmergencyNotificaton(self, time_interval):

        try:
            while True:
                price = self.getBitcoinPrice()
                value = {'value1': price}
                self.postWebhooks('bitcoin_price_emergency', value)
                print("Emergency notification sent")
                time.sleep(float(time_interval[0]) * 60)

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')

    def sendTelegramNotificaton(self, time_interval):

        try:
            while True:
                price = self.getBitcoinPrice()
                value = {'value1': price}
                self.postWebhooks('bitcoin_price_update', value)
                print("Telegram Notification sent")
                time.sleep(float(time_interval[0]) * 60)

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')

    def sendGmailNotification(self, time_interval):

        try:
            regex = r'^[a-z0-9]+[\._]?[a-z0-9]+[@]\w+[.]\w{2,3}$'
            name = input('Please enter your name: ')
            email = input('Please enter your Email: ')
            while True:

                # chceking email is valid
                if(re.search(regex, email)):
                    price = self.getBitcoinPrice()
                    value = {'value1': email, 'value2': name, 'value3': price}
                    self.postWebhooks('email_bitcoin_notification', value)
                    print('Email notification sent')
                    time.sleep(float(time_interval[0]) * 60)
                else:
                    print('Invalid Email ,' + name)
                    print('Please re-run the program and provide valid email')
                    sys.exit()

        except KeyboardInterrupt:
            print('Exiting from this application, Please wait.....')
            time.sleep(5)
            print('ThankYou for using this Application')


def main():

    Bitcoin_Api_URL = 'https://api.coindesk.com/v1/bpi/currentprice/INR.json'
    # for bitcoin.notificaion123@gmail.com
    url = 'https://maker.ifttt.com/trigger/{}/with/key/'
    key = 'jrvNIRNL-vboEF2g3mcaDoUitHgy-Z180veQ4JkbYNf'
    Webhooks_URL = url + key
    b1 = BitcoinNotification(Bitcoin_Api_URL, Webhooks_URL)
    current_bitcoinprice = b1.getBitcoinPrice()
    parser = argparse.ArgumentParser(
        usage='''\
            Usage: This app gives the price of 1 Bitcoin in INR.
            Destination(-d) must be provided else destination will be
            telegram by default.
            To recive notification from IFTTT install IFTTT mobile app.
            To recive notifications on Telegram install Telegram app
            and join this channel @IFTTT.

            Prerequisite : INSTALL IFTTT APP AND TELEGRAM APP
            and TO RECIVE NOTIFICATON.
            PRESS Ctrl+C to terminate the app
        ''',
        description="BITCOIN PRICE NOTIFIER ",
        epilog="Copyright @Rajat Bhatt")
    # command line variable for time gap
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        nargs=1,
        metavar="interval",
        default=[1],
        help="Time interval in minutes")
    # command line variable for threshold
    parser.add_argument(
        "-t",
        "--threshold",
        type=int,
        nargs=1,
        metavar="threshold",
        default=[400000],
        help="Threshold in INR(₹)")

    # command line variable for destination
    parser.add_argument(
        "-d",
        "--destination",
        metavar='destnation',
        default='telegram',
        help='There are various options to recive notifications'
        'from us (1)IFTTT app (2) Telegram app (3) Email')

    new_args = parser.parse_args()
    print('Running Application with time interval of ', new_args.interval[0],
          'mins and threshold = ₹', new_args.threshold[0], 'and Destination= ',
          new_args.destination)

    # calls the function to send notifications
    if (new_args.destination == 'telegram' and
            new_args.threshold[0] < current_bitcoinprice):
        print('''\
            To recive the notification
        from Telegram, install the telegram app and join the
        channel using this link - https://t.me/bitcoin_notificationrajat or
        follow @Bitcoinprice_notificationrajat
        ''')
        b1.sendTelegramNotificaton(new_args.interval)

    if (new_args.destination == 'ifttt' or
            new_args.threshold[0] > current_bitcoinprice):

        b1.sendIFTTTEmergencyNotificaton(new_args.interval)

    if (new_args.destination == 'email'):
        b1.sendGmailNotification(new_args.interval)


# driver code
if __name__ == "__main__":
    main()
